File: src/com/train/algorithm/search/implementInC++/LetterCasePermutation.cpp
//
// Created by 罗斯 on 12/9/2022.
//
#include <vector>
#include <string>
using namespace std;


class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> res;
        dfs(s,0,res);
        return res;
    }

private:
    void dfs(string& s,int i,vector<string>& res){
        if(i == s.length()){
            res.push_back(s);
            return;
        }

        dfs(s,i+1,res);
        if(!isalpha(s[i])) return;
        s[i] ^= (1 << 5);
        dfs(s,i+1,res);
        s[i] ^= (1 << 5);
    }
};

File: src/com/train/algorithm/search/implementInPython/LetterCasePermutation.py
class Solution(object):
    def letterCasePermutation(self, s):
        res = []

        def dfs(sub="", i=0):
            if len(sub) == len(s):
                res.append(sub[:])
                return
            for j in range(i, len(s)):
                if s[j].isdigit():
                    dfs(sub + s[j], j + 1)
                else:
                    dfs(sub + s[j].lower(), j + 1)
                    dfs(sub + s[j].upper(), j + 1)

        dfs()
        return res